frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSquashAction.java

```diff
@@ -0,0 +1,180 @@
+          .flatMap(cbn -> branchNameOption.map(bn -> bn.equals(cbn))).getOrElse(false);
+      doSquash(project, gitRepository, parent, commits, branchName, isSquashingCurrentBranch);
+    }
+  }
+
+  @UIEffect
+  private void doSquash(Project project,
+      GitRepository gitRepository,
+      ICommitOfManagedBranch parent,
+      List<ICommitOfManagedBranch> commits,
+      String branchName,
+      boolean isSquashingCurrentBranch) {
+
+    @Data
+    // So that Interning Checker doesn't complain about enum comparison (by `equals` and not by `==`) in Lombok-generated `equals`
+    @SuppressWarnings("interning:not.interned")
+    class VcsCommitMetadataAndMessage {
+      private final List<VcsCommitMetadata> metadata;
+      private final String message;
+
+    }
+
+    val vcsCommitMetadataAndMessage = commits.foldLeft(
+        new VcsCommitMetadataAndMessage(List.empty(), ""),
+        (acc, commit) -> new VcsCommitMetadataAndMessage(
+            acc.metadata.append(new VcsCommitMetadataAdapterForSquash(parent, commit)),
+            "${acc.message}${commit.getFullMessage()}${NL}${NL}"));
+
+    val dialog = new GitNewCommitMessageActionDialog(
+        /* project */ project,
+        /* message */ vcsCommitMetadataAndMessage.message,
+        /* title */ getNonHtmlString("action.GitMachete.BaseSquashAction.dialog.title"),
+        /* dialogLabel */ getNonHtmlString("action.GitMachete.BaseSquashAction.dialog.label"));
+
+    String taskName = isSquashingCurrentBranch
+        ? getString("action.GitMachete.BaseSquashAction.task-title.current")
+        : getString("action.GitMachete.BaseSquashAction.task-title.non-current");
+
+    dialog.show(
+        newMessage -> {
+          new Task.Backgroundable(project, taskName) {
+            @Override
+            @UIThreadUnsafe
+            public void run(ProgressIndicator indicator) {
+              LOG.info("Checking out '${branchName}' branch and squashing it");
+
+              if (!isSquashingCurrentBranch) {
+                CheckoutSelectedBranchAction.doCheckout(project, indicator, branchName, gitRepository);
+              }
+
+              val commitsToSquash = vcsCommitMetadataAndMessage.metadata.toJavaList();
+              val operationResult = new GitSquashOperation(gitRepository).execute(commitsToSquash, newMessage);
+
+              if (isComplete(operationResult)) {
+                val title = getString("action.GitMachete.BaseSquashAction.notification.title");
+                val notification = VcsNotifier.STANDARD_NOTIFICATION.createNotification(title, NotificationType.INFORMATION);
+                VcsNotifier.getInstance(project).notify(notification);
+              }
+            }
+          }.queue();
+
+          return Unit.INSTANCE;
+        });
+  }
+
+  /**
+   * Hackish approach to check for kotlin internal sealed class {@link git4idea.rebase.log.GitCommitEditingOperationResult.Complete}.
```

Comment:
Huh actually... pls check if this PR passes UI tests against earlier supported versions of IntelliJ (as such tests are only run by default on `develop` and `master` builds, not on PR builds for the sake of CI performance). A matter of changing `when` condition in CircleCI config

Reply:
(btw, is this line covered by UI tests? we don't have test coverage wired up in this project, alas... might be non-trivial for UI tests huh)

Reply:
(https://github.com/VirtusLab/git-machete-intellij-plugin/issues/1060)

Reply:
https://app.circleci.com/pipelines/github/VirtusLab/git-machete-intellij-plugin/6011/workflows/5aa221c1-62e0-4360-a276-38fa1b3ffd82/jobs/6341

Reply:
UI tests against earlier supported versions of IntelliJ ✅